src/cc/bcc_btf.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2019 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef BCC_BTF_H
+#define BCC_BTF_H
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string>
+#include <map>
+#include <vector>
+
+struct btf;
+struct btf_ext;
+
+namespace ebpf {
+
+class BTFStringTable {
+ private:
+  uint32_t Size;
+  std::map<uint32_t, uint32_t> OffsetToIdMap;
+  std::vector<std::string> Table;
+
+ public:
+  BTFStringTable(): Size(0) {}
+  uint32_t getSize() { return Size; }
+  std::vector<std::string> &getTable() { return Table; }
+  uint32_t addString(std::string Str);
+};
+
+class BTF {
+ public:
+  ~BTF();
+  int load(uint8_t *btf_sec, uintptr_t btf_sec_size,
+           uint8_t *btf_ext_sec, uintptr_t btf_ext_sec_size,
+           std::map<std::string, std::string> &remapped_sources);
+  int get_fd();
+  int get_btf_info(const char *fname,
+                   void **func_info, unsigned *func_info_cnt,
+                   unsigned *finfo_rec_size,
+                   void **line_info, unsigned *line_info_cnt,
+                   unsigned *linfo_rec_size);
+
+ private:
+  void adjust(uint8_t *btf_sec, uintptr_t btf_sec_size,
+              uint8_t *btf_ext_sec, uintptr_t btf_ext_sec_size,
+              std::map<std::string, std::string> &remapped_sources,
+              uint8_t **new_btf_sec, uintptr_t *new_btf_sec_size);
+
+ private:
+  struct btf *btf_;
```

Comment:
I think would have been cleaner to create pure c++ wrapper for btf_* where btf__new() would have been a constructor and btf__free() called in destructor. Another pure c++ wrapper for btf_ext__*(). Than this class would be called BCC_BTF, since it has this 'adjust()' workaround and private members at this line would be unique_ptr<BTF> and unique_ptr<BTF_EXT>.
May be such c++ wrappers should be in github/libbpf/ repo.
Such refactoring can be done later.
Overall great stuff!

Reply:
Agree. I think some extensible C++ wrapper will be good for common use. Implementing in libbpf repo certainly is a great way for reusability. As you suggested, will try to refactor later.